token/program-2022/src/processor.rs

```diff
@@ -451,6 +451,9 @@ impl Processor {
         // This check MUST occur just before the amounts are manipulated
         // to ensure self-transfers are fully validated
         if self_transfer {
+            if memo_required(&source_account) {
+                check_previous_sibling_instruction_is_memo()?;
+            }
```

Comment:
Does it maybe make sense to just move the existing check from L469 before this block so we don't have to duplicate?
https://github.com/solana-labs/solana-program-library/blob/d103601f6f83d26e30189959763da1c894ec1d44/token/program-2022/src/processor.rs#L453-L455

Reply:
We can't simply move the check up unfortunately. We borrowing the destination account data as mutable to manipulate its data and check for the memo requirement, but if it's a self-transfer, we do a double-mutable borrow between the source and destination account data and panic.

I couldn't think of a better way to get the check done unfortunately than to duplicate the check on the source account. Do you have any suggestions?

Reply:
Ugh, that's annoying. I despise the brittleness of having 2 memo checks in different spots.
But I haven't come up with anything better so far. 😞